### PR TITLE
post about docker params change detection

### PR DESCRIPTION
**Pull request by adrianiurca**

### CODE REVIEW

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  - development
+tags:
+  - bugfix
+  - docker
+  - personal
+---
+
+## Docker params change detection
+
+An interesting behavior was present in docker::run component. The problem was that if at least one parameter was added/modified/removed puppet agent didn't apply the change only if you stopped, removed manually the container, and reapply the manifest. The solution was to create a new function that detects if at least one parameter is changed. The detection mechanism is based on the check between parameter values from the manifest file and correspondent field from the docker inspect object of the currently running container. The solution is present in the 3.11.0 version. So now let's see how the problem can be reproduced by using version <= 3.10.2:
```
Comment (carabasdaniel) [inline _posts/2020-09-18-docker-params-change-detection.md:16]:
```suggestion
An interesting behaviour was present in docker::run component. The problem was that if any parameter was added/modified/removed puppet agent didn't apply the change only if you stopped, removed manually the container, and reapply the manifest, forcing a new container creation. 

The solution was to create a new function that detects if at least one parameter is changed. The detection mechanism is based on the check between parameter values from the manifest file and correspondent field from the docker inspect object of the currently running container. The solution is present in the 3.11.0 version. 

So now let's see how the problem can be reproduced by using version <= 3.10.2:
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+$: hello-world:linux
+```
+
+so... something is wrong, the tag was not changed to `hello-world:latest`. If we want to apply this change we need to do a few more steps:
+
+- stop the container: `docker stop servercore`
+- remove the container: `docker rm servercore`
+- reapply the manifest detailed above: `puppet apply <manifest_file_name>`
+
+In conclusion in the puppetlabs/docker module versions <= 3.10.2 the parameter change is not detected. If we want to change some parameters for the same container, the puppet agent will not apply these changes for us until we delete the container manually. If we use the latest version(3.11.0) this problem is resolved. The parameter detection mechanism is implemented for the most important parameters such as image, volumes and ports, for the rest of the parameters new PRs are always welcomed.
```
Comment (carabasdaniel) [inline _posts/2020-09-18-docker-params-change-detection.md:75]:
```suggestion
In conclusion in the puppetlabs/docker module versions <= 3.10.2 the parameter change is not detected. If we want to change some parameters for the same container, the puppet agent will not apply these changes for us until we delete the container manually. 
Using the latest versions(>=3.11.0) this problem is resolved by having the parameter detection mechanism implemented for the most important parameters such as image, volumes and ports.
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  - development
+tags:
+  - bugfix
+  - docker
+  - personal
+---
+
+## Docker params change detection
+
+An interesting behaviour was present in docker::run component. The problem was that if any parameter was added/modified/removed puppet agent didn't apply the change only if you stopped, removed manually the container, and reapply the manifest, forcing a new container creation.
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:16]:
```suggestion
An interesting behaviour was present in docker::run component. The problem was that if any parameter was added/modified/removed puppet agent would only apply the change only if you stopped, removed the container manually and reapplied the manifest, forcing a new container creation. 
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  - bugfix
+  - docker
+  - personal
+---
+
+## Docker params change detection
+
+An interesting behaviour was present in docker::run component. The problem was that if any parameter was added/modified/removed puppet agent didn't apply the change only if you stopped, removed manually the container, and reapply the manifest, forcing a new container creation. 
+
+The solution was to create a new function that detects if at least one parameter is changed. The detection mechanism is based on the check between parameter values from the manifest file and correspondent field from the docker inspect object of the currently running container. The solution is present in the 3.11.0 version.
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:18]:
```suggestion
The solution was to create a new function that detects if at least one parameter is changed. The detection mechanism is based on the check between parameter values from the manifest file and correspondent field from the docker inspect object of the currently running container. The solution is present in the `3.11.0` version. 
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  - personal
+---
+
+## Docker params change detection
+
+An interesting behaviour was present in docker::run component. The problem was that if any parameter was added/modified/removed puppet agent didn't apply the change only if you stopped, removed manually the container, and reapply the manifest, forcing a new container creation. 
+
+The solution was to create a new function that detects if at least one parameter is changed. The detection mechanism is based on the check between parameter values from the manifest file and correspondent field from the docker inspect object of the currently running container. The solution is present in the 3.11.0 version. 
+
+So now let's see how the problem can be reproduced by using version <= 3.10.2:
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:20]:
```suggestion
So now let's see how the problem can be reproduced by using version `<= 3.10.2`:
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+$: hello-world:linux
+```
+
+so... something is wrong, the tag was not changed to `hello-world:latest`. If we want to apply this change we need to do a few more steps:
+
+- stop the container: `docker stop servercore`
+- remove the container: `docker rm servercore`
+- reapply the manifest detailed above: `puppet apply <manifest_file_name>`
+
+In conclusion in the puppetlabs/docker module versions <= 3.10.2 the parameter change is not detected. If we want to change some parameters for the same container, the puppet agent will not apply these changes for us until we delete the container manually.
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:79]:
```suggestion
In conclusion in the puppetlabs/docker module versions `<= 3.10.2 `the parameter change is not detected. If we want to change some parameters for the same container, the puppet agent will not apply these changes for us until we delete the container manually. 
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+so... something is wrong, the tag was not changed to `hello-world:latest`. If we want to apply this change we need to do a few more steps:
+
+- stop the container: `docker stop servercore`
+- remove the container: `docker rm servercore`
+- reapply the manifest detailed above: `puppet apply <manifest_file_name>`
+
+In conclusion in the puppetlabs/docker module versions <= 3.10.2 the parameter change is not detected. If we want to change some parameters for the same container, the puppet agent will not apply these changes for us until we delete the container manually. 
+Using the latest versions(>=3.11.0) this problem is resolved by having the parameter detection mechanism implemented for the most important parameters such as image, volumes and ports.
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:80]:
```suggestion
Using the latest versions(`>=3.11.0`) this problem is resolved by having the parameter detection mechanism implemented for the most important parameters such as image, volumes and ports.
```

File: _posts/2020-09-18-docker-params-change-detection.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+Also please take a look at the [sollution](https://github.com/puppetlabs/puppetlabs-docker/pull/648).
```
Comment (sanfrancrisko) [inline _posts/2020-09-18-docker-params-change-detection.md:81]:
```suggestion
Also please take a look at the [solution](https://github.com/puppetlabs/puppetlabs-docker/pull/648).
```

**Review by sanfrancrisko — COMMENTED** [review]
Great post @adrianiurca !

Just a few grammatical and formatting suggestions, but otherwise LGTM! 😄

**Comment by DavidS** [comment]
Unless this is urgent to go out, please schedule it for wednesday week (by setting the date in the file name to that date). That'll will spread out readers a bit and is generally supposed to improve engagement (nobody wants to read two posts on the same day from the same team, even if they're good posts.)

File: _posts/2020-09-18-docker-params-change-detection.md (hunk 4 above)
Comment (adrianiurca) [reply _posts/2020-09-18-docker-params-change-detection.md]: I commited your changes. Thank you @sanfrancrisko

**Review by sanfrancrisko — COMMENTED** [review]
LGTM, but I like @DavidS 's [suggestion](https://github.com/puppetlabs/iac/pull/111#issuecomment-694753770) of landing this next week some time. If you change the page name to `2020-09-23` we can merge it today and then it auto publishes next Wed.

**Review by sanfrancrisko — APPROVED** [review]

**Pull request merged**